include/server/mir/frontend/surface_stack.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2019 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_FRONTEND_SURFACE_STACK_H_
+#define MIR_FRONTEND_SURFACE_STACK_H_
+
+#include <memory>
+
+namespace mir
+{
+namespace scene
+{
+class Observer;
+}
+namespace frontend
+{
+class SurfaceStack
+{
+public:
+    SurfaceStack() = default;
+    virtual ~SurfaceStack() = default;
+
+    virtual void add_observer(std::shared_ptr<scene::Observer> const& observer) = 0;
+    virtual void remove_observer(std::weak_ptr<scene::Observer> const& observer) = 0;
```

Comment:
I don't think we'll make enough use of this to be worth suggesting that `scene::Observer` is too wide an interface for the frontend requirements.

\<rant>
C++ really doesn't have good support for designs where `frontend::SceneObserver` only offers some of the `scene::Observer` methods.

E.g. `frontend::SceneObserver` could no-op & `final` a bunch of unwanted stuff. But there's no contravariant  parameter passing for `std::shared_ptr<>` that allows it to be passed directly. (That would need an addition forwarding function here to force the conversion.)
\</rant>